src/rnnlm/rnnlm-simple-looped.h

```diff
@@ -0,0 +1,184 @@
+
+  const kaldi::nnet3::Nnet &rnnlm;
+  const CuMatrix<BaseFloat> &word_embedding_mat;
+
+  // frames_left_context equals the model left context plus the value of the
+  // --extra-left-context-initial option.
+  int32 frames_left_context;
+  // frames_right_context is the same as the right-context of the model.
+  int32 frames_right_context;
+  // The frames_per_chunk equals the number of input frames we need for each
+  // chunk (except for the first chunk).
+  int32 frames_per_chunk;
+
+  // The output dimension of the nnet neural network (not the final output).
+  int32 nnet_output_dim;
+
+  // The 3 computation requests that are used to create the looped
+  // computation are stored in the class, as we need them to work out
+  // exactly shich iVectors are needed.
+  ComputationRequest request1, request2, request3;
+
+  // The compiled, 'looped' computation.
+  NnetComputation computation;
+};
+
+/*
+  This class handles the neural net computation; it's mostly accessed
+  via other wrapper classes.
+
+  It accept just input features */
+class RnnlmSimpleLooped {
+ public:
+  /**
+     This constructor takes features as input.
+     Note: it stores references to all arguments to the constructor, so don't
+     delete them till this goes out of scope.
+
+     @param [in] info   This helper class contains all the static pre-computed information
+                        this class needs, and contains a pointer to the neural net.
+     @param [in] feats  The input feature matrix.
+  */
+  RnnlmSimpleLooped(const RnnlmSimpleLoopedInfo &info);
+
+  RnnlmSimpleLooped(const RnnlmSimpleLooped &other);
+
+  inline int32 NnetOutputDim() const { return info_.nnet_output_dim; }
+
+  // Gets the nnet's output for a particular frame, with 0 <= frame < NumFrames().
+  // 'output' must be correctly sized (with dimension NnetOutputDim()).  Note:
+  // you're expected to call this, and GetOutput(), in an order of increasing
+  // frames.  If you deviate from this, one of these calls may crash.
+//  void GetNnetOutputForFrame(int32 frame, VectorBase<BaseFloat> *output);
+
+  // Updates feats_ with the new incoming word specified in word_indexes
+  // We usually do this one at a time
+  void TakeFeatures(const std::vector<int32> &word_indexes);
+
+  // Gets the output for a particular frame and word_index, with
+  // 0 <= frame < NumFrames().
+//  BaseFloat GetOutput(int32 frame, int32 word_index);
+  // create a CuVector in heap, pointer owned by the caller
+  CuVector<BaseFloat>* GetOutput(int32 frame);
+
+  BaseFloat LogProbOfWord(int32 word_index,
+                          const CuVectorBase<BaseFloat> &hidden) const;
+
+ private:
+  // This function does the computation for the next chunk.
+  void AdvanceChunk();
+
+  const RnnlmSimpleLoopedInfo &info_;
+
+  NnetComputer computer_;
+
+  SparseMatrix<BaseFloat> feats_;
+
+  // The current nnet's output that we got from the last time we
+  // ran the computation.
+  Matrix<BaseFloat> current_nnet_output_;
```

Comment:
... and there no need to store the nnet output in CPU memory.  That was done in the decoding case because single-element lookups are slow to do one by one to GPU.  But if we are using a GPU, we'll want to do the vector dot-products on GPU.